### ISSUE TITLE
Make auxiliary dependencies optional

### ISSUE THREAD

**Issue by kiudee** [kiudee/cs-ranking#67]
Currently we require quite a few dependencies, which makes installing the library difficult. I went ahead and categorized the different dependencies (see below). We should do the following:

1. Remove these dependencies from `install_requires`
2. Move these to [`extras_require`](https://setuptools.readthedocs.io/en/latest/setuptools.html#declaring-extras-optional-features-with-their-own-dependencies) 

Dependencies
------------

- Core (mandatory):
  - numpy
  - scipy
  - scikit-learn
  - scikit-optimize
  - joblib
  - keras
  - tensorflow
  - docopt
- Data I/O or generation (optional):
  - psycopg2-binary for database access
  - pandas
  - h5py
  - pygmo
- Required for some of the probabilistic models (optional):
  - pymc3
  - theano
- Nice to haves (optional):
  - tqdm for progress bars

**Comment by timokau** [comment]
We will also need to come up with a way to [gracefully handle](https://stackoverflow.com/questions/15928101/python-how-to-check-extra-requirements-at-runtime) the case where those dependencies are not available. A runtime warning is probably good enough here.

**Comment by timokau** [comment]
Another thing we need to think about is how to actually add the extra requirements. Currently we add requirements to `requirements.txt` and then parse that file into `setup.py` as well. That gives the advantage of having the requirements properly declared for `setuptools` while still being able to just use `pip install -r requirements.txt`, e.g. in a virtualenv.

The problem is that extras are not [supported](https://github.com/pypa/pip/issues/1161) in the `requirements.txt` file. So we could

1. Declare all requirements in `requirements.txt` and hack around with some comment syntax that marks blocks that belong to extras (like everything after `# [some-extra]`). That would not give very much flexibility when using `requirements.txt`, but most of the usecases regarding that file are probably development anyways. In a development context, installing everything is probably the right call in most cases.

	Downsides: hacks needed, not much flexibility
	Upsides: no extra file, "just works" transparently
2. Create a separate `requirements-<some-extra>.txt` file for every extra.

	Downsides: lots of files, annoying to set up an environment that includes everything
	Upsides: no hacks, maximum flexibility
3. Do not declare the extra requirements in `requirements.txt` at all, only in `setup.py`.

	Downsides: need to be installed manually in dev environments
	Upsides: very simple, no hacks

I think I'm partial to either (1) or (3).

**Comment by kiudee** [comment]
I think I prefer (3), moving everything into `setup.py`. This is also what I am doing in all my other projects. A separate requirements-dev.txt can still useful for requirements only useful for testing etc.


Regarding the graceful handling of dependencies: Another "solution" I have seen is to import the relevant dependencies inside the actual functions/classes, where they are used. I prefer having all the dependencies specified in the header of the `.py` file, which is why I think the try/except is a better solution.

**Comment by timokau** [comment]
I have started with this work in https://github.com/kiudee/cs-ranking/pull/90. What do you think about the approach?

**Comment by timokau** [comment]
Implemented in #90.
